sql/core/src/main/scala/org/apache/spark/sql/Observation.scala

```diff
@@ -0,0 +1,147 @@
+package org.apache.spark.sql
+
+import java.util.UUID
+import java.util.concurrent.TimeUnit
+import java.util.concurrent.locks.{Condition, Lock, ReentrantLock}
+
+import org.apache.spark.sql.execution.QueryExecution
+import org.apache.spark.sql.util.QueryExecutionListener
+
+/**
+ * Not thread-safe.
+ * @param name
+ * @param sparkSession
+ */
+class Observation(name: String) {
+
+  private val lock: Lock = new ReentrantLock()
+  private val completed: Condition = lock.newCondition()
+  private val listener: ObservationListener = ObservationListener(this)
+
+  private var sparkSession: Option[SparkSession] = None
+
+  @transient private var row: Option[Row] = None
+
+  /**
+   * Attach this observation to the given Dataset.
+   * Remember to call `close()` when the observation is done.
+   *
+   * @param ds dataset
+   * @tparam T dataset type
+   * @return observed dataset
+   */
+  def on[T](ds: Dataset[T])(expr: Column, exprs: Column*): Dataset[T] = {
+    if (ds.isStreaming) {
+      throw new IllegalArgumentException("Observation does not support streaming Datasets")
+    }
+    register(ds.sparkSession)
+    ds.observe(name, expr, exprs: _*)
+  }
+
+  /**
+   * Wait for the first action on the observed dataset to complete and returns true.
+   * This method times out after the given amount of time and returns false.
+   *
+   * @param time timeout
+   * @param unit timeout time unit
+   * @return true if action complete within timeout, false on timeout
+   */
+  def waitCompleted(time: Long, unit: TimeUnit): Boolean = waitCompleted(Some(time), unit)
+
+  /**
+   * Get the observation results. This waits until the observed dataset finishes its first action.
+   * If you want to wait for the result and provide a timeout, use waitCompleted.
+   * Only the result of the first action is available. Subsequent actions do not modify the result.
+   */
+  def get: Row = {
+    waitCompleted(None, TimeUnit.SECONDS)
+    row.get
+  }
+
+  private def waitCompleted(time: Option[Long], unit: TimeUnit): Boolean = {
+    lock.lock()
+    try {
+      if (row.isEmpty) {
+        if (time.isDefined) {
+          completed.await(time.get, unit)
+        } else {
+          completed.await()
+        }
+      }
+      row.isDefined
+    } finally {
+      lock.unlock()
+    }
+  }
+
+  private def getMetricRow(metrics: Map[String, Row]): Option[Row] =
+    metrics
+      .find { case (metricName, _) => metricName.equals(name) }
+      .map { case (_, row) => row }
+
+  private def register(sparkSession: SparkSession): Unit = {
+    if (this.sparkSession.isDefined) {
+      throw new IllegalStateException("An Observation can be used with a Dataset only once")
+    }
+    this.sparkSession = Some(sparkSession)
+    sparkSession.listenerManager.register(listener)
+  }
+
+  private def unregister(): Unit = {
+    this.sparkSession.foreach(_.listenerManager.unregister(listener))
+  }
+
+  private[spark] def onFinish(funcName: String, qe: QueryExecution): Unit = {
```

Comment:
This is called by the listener, which should always be a single thread, seems no need to lock.

Reply:
But the listener is called by a different thread than the one calling into `waitCompleted`, right? We are transferring data from one thread (the one calling into `onFinish`) to the main thread waiting for the observation result calling into `get` / `waitCompleted`. This needs synchronization, otherwise the main threads will not see the result if it is faster than the listener thread.